uspace/lib/usbhost/include/usb/host/utility.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2017 Ondrej Hlavaty <aearsis@eideo.cz>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * - Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in the
+ *   documentation and/or other materials provided with the distribution.
+ * - The name of the author may not be used to endorse or promote products
+ *   derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/** @addtogroup drvusbehci
+ * @{
+ */
+/** @file
+ * @brief USB host controller library: Utility functions.
+ *
+ * A mix of various functions that makes life of USB HC driver developers
+ * easier.
+ */
+#ifndef LIB_USBHOST_UTILITY
+#define LIB_USBHOST_UTILITY
+
+#include <usb/host/bus.h>
+#include <usb/host/usb_transfer_batch.h>
+#include <usb/descriptor.h>
+#include <usb/classes/hub.h>
+#include <usb/request.h>
+
+typedef void (*endpoint_reset_toggle_t)(endpoint_t *);
+
+uint16_t hc_get_ep0_initial_mps(usb_speed_t);
+int hc_get_ep0_max_packet_size(uint16_t *, device_t *);
+void hc_reset_toggles(const usb_transfer_batch_t *batch, endpoint_reset_toggle_t);
+int hc_setup_virtual_root_hub(hc_device_t *, usb_speed_t);
+int hc_get_device_desc(device_t *, usb_standard_device_descriptor_t *);
+int hc_get_hub_desc(device_t *, usb_hub_descriptor_header_t *);
+int hc_device_explore(device_t *);
+
+/** Joinable fibril */
```

Comment:
This should be imo part of the stock fibril support in libc.

Reply:
Ideally by making fibril joinable by default.

Reply:
Yeah, we think so. But we like to introduce such changes in our subtree first, they can be moved easily later.

Reply:
I see, but there is really no benefit in hiding this in your subtree when you have your own branch. See the note about self-contained code here: http://www.helenos.org/wiki/StudentTips#Implementation

Reply:
OK, let's move it to libc. I don't think we shall change the behavior of fibril (making fid_t a structure with former fid_t, mutex, condvar etc.). This is more an "async task", though the term "task" shall probably not be used to avoid confusion. How do you think it should be called?

Reply:
The only way how to make fibrils joinable is making `fibril_t` reference-counted. If that is feasible, we can make fibrils joinable on-demand before starting them. The current usage would be preserved, and code that would want to join their fibrils would have to increase the refcount, join and decrease the refcount again. (This has even the neat imaginary benefit of allowing more fibrils to join on one :) ). Counterexamples, opinions, comments, acks?

Reply:
I think the general concensus the last time we discussed fibrils is that we want to move towards a more generic threading API, and deprecate any use of fibrils that doesn't work with the standard APIs. That means we should go with the traditional `join()`/`detach()` scheme, to avoid future headaches. We'll probably also need a new creation function, and leave the current one as detached-by-default, to avoid changes to existing code. I thing we'll eventually drop the fibril API entirely, in favor of C11's <threads.h>, so there's no point in fussing with it too much.

Reply:
In other words, essentialy reference counting with default value 2, and fibril_create calling fibril_create_joinable and dropping the reference? That could work. Shall I prepare it?

Reply:
Yeah, something like that. Please do, thanks! :)